//
// This file was generated by the JavaTM Architecture for XML Binding(JAXB) Reference Implementation, v2.2.8-b130911.1802 
// See <a href="http://java.sun.com/xml/jaxb">http://java.sun.com/xml/jaxb</a> 
// Any modifications to this file will be lost upon recompilation of the source schema. 
// Generated on: 2020.02.28 at 02:28:17 PM CST 
//


package utsw.bicf.answer.model.ehr;

import java.util.ArrayList;
import java.util.List;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElements;
import javax.xml.bind.annotation.XmlType;


/**
 * <p>Java class for ShortVariant complex type.
 * 
 * <p>The following schema fragment specifies the expected content contained within this class.
 * 
 * <pre>
 * &lt;complexType name="ShortVariant">
 *   &lt;complexContent>
 *     &lt;extension base="{http://foundationmedicine.com/compbio/variant-report-external}Variant">
 *       &lt;choice maxOccurs="unbounded" minOccurs="0">
 *         &lt;element name="dna-evidence" type="{http://foundationmedicine.com/compbio/variant-report-external}ShortVariantDnaEvidence"/>
 *         &lt;element name="rna-evidence" type="{http://foundationmedicine.com/compbio/variant-report-external}ShortVariantRnaEvidence"/>
 *       &lt;/choice>
 *       &lt;attribute name="cds-effect" use="required" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       &lt;attribute name="depth" use="required" type="{http://www.w3.org/2001/XMLSchema}int" />
 *       &lt;attribute name="gene" use="required" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       &lt;attribute name="percent-reads" use="required" type="{http://foundationmedicine.com/compbio/variant-report-external}Percent" />
 *       &lt;attribute name="allele-fraction" use="required" type="{http://foundationmedicine.com/compbio/variant-report-external}Fraction" />
 *       &lt;attribute name="position" use="required" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       &lt;attribute name="protein-effect" use="required" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       &lt;attribute name="transcript" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       &lt;attribute name="strand" type="{http://foundationmedicine.com/compbio/variant-report-external}Strand" />
 *       &lt;attribute name="functional-effect" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       &lt;attribute name="subclonal" type="{http://www.w3.org/2001/XMLSchema}boolean" default="false" />
 *     &lt;/extension>
 *   &lt;/complexContent>
 * &lt;/complexType>
 * </pre>
 * 
 * 
 */
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(name = "ShortVariant", propOrder = {
    "dnaEvidenceOrRnaEvidence"
})
public class ShortVariant
    extends Variant
{

    @XmlElements({
        @XmlElement(name = "dna-evidence", type = ShortVariantDnaEvidence.class),
        @XmlElement(name = "rna-evidence", type = ShortVariantRnaEvidence.class)
    })
    protected List<VariantEvidence> dnaEvidenceOrRnaEvidence;
    @XmlAttribute(name = "cds-effect", required = true)
    protected String cdsEffect;
    @XmlAttribute(name = "depth", required = true)
    protected int depth;
    @XmlAttribute(name = "gene", required = true)
    protected String gene;
    @XmlAttribute(name = "percent-reads", required = true)
    protected float percentReads;
    @XmlAttribute(name = "allele-fraction", required = true)
    protected float alleleFraction;
    @XmlAttribute(name = "position", required = true)
    protected String position;
    @XmlAttribute(name = "protein-effect", required = true)
    protected String proteinEffect;
    @XmlAttribute(name = "transcript")
    protected String transcript;
    @XmlAttribute(name = "strand")
    protected String strand;
    @XmlAttribute(name = "functional-effect")
    protected String functionalEffect;
    @XmlAttribute(name = "subclonal")
    protected Boolean subclonal;

    /**
     * Gets the value of the dnaEvidenceOrRnaEvidence property.
     * 
     * <p>
     * This accessor method returns a reference to the live list,
     * not a snapshot. Therefore any modification you make to the
     * returned list will be present inside the JAXB object.
     * This is why there is not a <CODE>set</CODE> method for the dnaEvidenceOrRnaEvidence property.
     * 
     * <p>
     * For example, to add a new item, do as follows:
     * <pre>
     *    getDnaEvidenceOrRnaEvidence().add(newItem);
     * </pre>
     * 
     * 
     * <p>
     * Objects of the following type(s) are allowed in the list
     * {@link ShortVariantDnaEvidence }
     * {@link ShortVariantRnaEvidence }
     * 
     * 
     */
    public List<VariantEvidence> getDnaEvidenceOrRnaEvidence() {
        if (dnaEvidenceOrRnaEvidence == null) {
            dnaEvidenceOrRnaEvidence = new ArrayList<VariantEvidence>();
        }
        return this.dnaEvidenceOrRnaEvidence;
    }

    /**
     * Gets the value of the cdsEffect property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getCdsEffect() {
        return cdsEffect;
    }

    /**
     * Sets the value of the cdsEffect property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setCdsEffect(String value) {
        this.cdsEffect = value;
    }

    /**
     * Gets the value of the depth property.
     * 
     */
    public int getDepth() {
        return depth;
    }

    /**
     * Sets the value of the depth property.
     * 
     */
    public void setDepth(int value) {
        this.depth = value;
    }

    /**
     * Gets the value of the gene property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getGene() {
        return gene;
    }

    /**
     * Sets the value of the gene property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setGene(String value) {
        this.gene = value;
    }

    /**
     * Gets the value of the percentReads property.
     * 
     */
    public float getPercentReads() {
        return percentReads;
    }

    /**
     * Sets the value of the percentReads property.
     * 
     */
    public void setPercentReads(float value) {
        this.percentReads = value;
    }

    /**
     * Gets the value of the alleleFraction property.
     * 
     */
    public float getAlleleFraction() {
        return alleleFraction;
    }

    /**
     * Sets the value of the alleleFraction property.
     * 
     */
    public void setAlleleFraction(float value) {
        this.alleleFraction = value;
    }

    /**
     * Gets the value of the position property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getPosition() {
        return position;
    }

    /**
     * Sets the value of the position property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setPosition(String value) {
        this.position = value;
    }

    /**
     * Gets the value of the proteinEffect property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getProteinEffect() {
        return proteinEffect;
    }

    /**
     * Sets the value of the proteinEffect property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setProteinEffect(String value) {
        this.proteinEffect = value;
    }

    /**
     * Gets the value of the transcript property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getTranscript() {
        return transcript;
    }

    /**
     * Sets the value of the transcript property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setTranscript(String value) {
        this.transcript = value;
    }

    /**
     * Gets the value of the strand property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getStrand() {
        return strand;
    }

    /**
     * Sets the value of the strand property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setStrand(String value) {
        this.strand = value;
    }

    /**
     * Gets the value of the functionalEffect property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getFunctionalEffect() {
        return functionalEffect;
    }

    /**
     * Sets the value of the functionalEffect property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setFunctionalEffect(String value) {
        this.functionalEffect = value;
    }

    /**
     * Gets the value of the subclonal property.
     * 
     * @return
     *     possible object is
     *     {@link Boolean }
     *     
     */
    public boolean isSubclonal() {
        if (subclonal == null) {
            return false;
        } else {
            return subclonal;
        }
    }

    /**
     * Sets the value of the subclonal property.
     * 
     * @param value
     *     allowed object is
     *     {@link Boolean }
     *     
     */
    public void setSubclonal(Boolean value) {
        this.subclonal = value;
    }

}
